torch/csrc/jit/script/compiler.cpp

```diff
@@ -106,14 +249,111 @@ struct to_ir {
+      return e.value_table.at(name);
+    }
+    return nullptr;
+  }
+
+  Value* findInParentFrame(const std::string& name) {
+    for (auto runner = environment_stack->next; runner; runner = runner->next) {
+      if (runner->value_table.count(name)) {
+        return runner->value_table.at(name);
+      }
+    }
+    return nullptr;
+  }
+
+  Value* createBlockInput(const std::string& name) {
+    auto& curr_frame = *environment_stack;
+    Block* b = curr_frame.b;
+
+    // Create the input
+    Value* new_input = b->addInput();
+
+    // Associate this name with this value
+    curr_frame.value_table[name] = new_input;
+
+    // List as a positional input
+    curr_frame.positional_inputs.push_back(name);
+
+    return new_input;
+  }
+
   void setVar(const std::string& name, Value* value) {
-    value_table[name] = value;
+    auto& curr_frame = *environment_stack;
+
+    Symbol owning_kind = Symbol();
+    if (curr_frame.b) {
+      owning_kind = curr_frame.b->owningNode()->kind();
+    }
+
+    if (owning_kind == Symbol("Loop") || owning_kind == Symbol("If")) {
+      // Overwriting an existing value means it's already been
+      // accounted for.
+      if (findInThisFrame(name)) {
+        curr_frame.value_table[name] = value;
+        return;
+      }
+
+      // Value not here. search in parent scopes
+      if (findInParentFrame(name)) {
+        // Writing to a value in a parent frame. Make this a
+        // loop-carried dependency.
+        if (owning_kind == Symbol("Loop"))
+          createBlockInput(name);
+
+        // Overwrite in value map
+        curr_frame.value_table[name] = value;
+      } else {
+        // Not accessing a value in enclosing scope. Make a new
+        // value as usual
+        curr_frame.value_table[name] = value;
+      }
+    } else {
+      curr_frame.value_table[name] = value;
+    }
   }
 
   Value* getVar(const Ident& ident) {
-    if (value_table.count(ident.name()) == 0)
-      throw ErrorReport(ident) << "undefined value " << ident.name();
-    return value_table[ident.name()];
+    try {
+      return getVar(ident.name());
+    } catch (ErrorReport e) {
+      throw ErrorReport(ident) << e.what();
+    }
+  }
+
+  Value* getVar(const std::string& ident) {
+    Value* retval = findInThisFrame(ident);
+    if (retval) {
+      return retval;
+    }
+
+    retval = findInParentFrame(ident);
+
+    auto& curr_frame = *environment_stack;
+
+    Symbol owning_kind = Symbol();
+    if (curr_frame.b) {
+      owning_kind = curr_frame.b->owningNode()->kind();
+    }
+
+    if (owning_kind == Symbol("Loop") || owning_kind == Symbol("If")) {
+      if (retval) {
+        // Reading from a value in a parent frame. Make this a
+        // loop-carried dependency or explicit input
+        if (owning_kind == Symbol("Loop"))
+          retval = createBlockInput(ident);
+
+        return retval;
+      } else {
+        throw ErrorReport() << "undefined value " << ident;
```

Comment:
This kinda stuff makes debugging hard. Can you pass a TreeRef anchor to each getVar call as a second argument (in the case of an Ident that is just the ident itself.)

Reply:
The above try-catch is supposed to catch this and throw a new one with `ident`. I was trying to factor this out to make it so that I can getVar without actually having an Ident available